Update middleware upgrade guide w/ more rationale (#38074)
vercel/next.js: errors/middleware-upgrade-guide.md

As we work on improving Middleware for General Availability (GA), we've made some changes to the Middleware APIs (and how you define Middleware in your application) based on your feedback.

This upgrade guide will help you understand the changes, why they were made, and how to migrate your existing Middleware to the new API. The guide is for Next.js developers who:

- Currently use the beta Next.js Middleware features
- Choose to upgrade to the next stable version of Next.js (`v12.2`)

You can start upgrading your Middleware usage today with the latest canary release (`npm i next@canary`).

> **Note**: these changes described in this guide are included in Next.js 12.2. You can keep your current site structure, including nested middleware until you move to 12.2 (or a canary build of Next.js).

Reviewer: ```suggestion
> **Note**: These changes described in this guide are included in Next.js `12.2`. You can keep your current site structure, including nested Middleware, until you move to `12.2` (or a `canary` build of Next.js).
```

> If you have ESLint configured, you will need to run `npm i eslint-config-next@canary --save-dev` to upgrade your ESLint configuration to ensure the same version is being used as the Next.js version. You might also need to restart VSCode for the changes to take effect.

Reviewer 2: ```suggestion
If you have ESLint configured, you will need to run `npm i eslint-config-next@latest --save-dev` to upgrade your ESLint configuration to ensure the same version is being used as the Next.js version. You might also need to restart VSCode for the changes to take effect.
```

Based on customer feedback, we have replaced this API with a single root Middleware, which provides the following improvements:

- **Faster execution with lower latency**: With nested Middleware, a single request could invoke multiple Middleware functions. A single Middleware means a single function execution, which is more efficient.
- **Less expensive**: Middleware usage is billed per invocation. Using nested Middleware, a single request could invoke multiple Middleware functions, meaning multiple Middleware charges per request. A single Middleware means a single invocation per request and is more cost effective.
- **Middleware can conveniently filter on things besides routes**: With nested Middleware, the Middleware files were located in the `pages` directory and Middleware was executed based on request paths. By moving to a single root Middleware, you can still execute code based on request paths, but you can now more conveniently execute Middleware based on other conditions, like `cookies` or the presence of a request header.

> - **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is harder to reason about. Moving to a single, root Middleware more explicitly defines execution order.

Reviewer 3: ```suggestion
- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is difficult to reason about. Moving to a single, root Middleware more explicitly defines execution order.
```

https://github.com/vercel/next.js/runs/7082368888?check_suite_focus=true#step:6:251

> - **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is not easy to reason about. Moving to a single, root Middleware more explicitly defines execution order.

Reviewer 3: ```suggestion
- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is harder to reason about. Moving to a single, root Middleware more explicitly defines execution order.
```

> - **Supports Next.js Layouts (RFC)**: Moving to a single, root mMddleware helps support the new [Layouts (RFC) in Next.js](https://nextjs.org/blog/layouts-rfc).

Reviewer 2: ```suggestion
- **Supports Next.js Layouts (RFC)**: Moving to a single, root Mddleware helps support the new [Layouts (RFC) in Next.js](https://nextjs.org/blog/layouts-rfc).
```

> - **Faster execution with lower latency**: with nested middleware, a single request could invoke multiple middleware functions. A single middleware means a single function execution, which is more efficient.
> - **Less expensive**: Middleware usage is billed per-invocation. Using nested middleware, a single request could invoke multiple middleware functions, meaning multiple middleware charges per request. A single middleware means a single invocation per request, and ultimately a single charge.
> - **Middleware can conveniently filter on things besides routes**: with nested middleware, the middleware files were located in the `pages` directory, and middleware was executed based on request paths. By moving to a single, root middleware, you can still execute certain code based on request path, but you can more conveniently execute middleware based on other conditions, like cookies or the presence of a request header.
> - **Straighforward, deterministic execution ordering**: with nested middleware, a single request could match multiple middleware functions. For example, a requests to `/dashboard/users/*` would invoke middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`, however the execution order is not straightforward to reason about. Moving to a single, root middleware puts execution order in the hand of the developer, so they can implement the order that makes the most sense for any given project.
> - **Supports Next.js Layouts (RFC)**: moving to a single, root middleware helps support [Layouts (RFC) in Next.js](https://nextjs.org/blog/layouts-rfc). The Layouts RFC proposes to move away from the current `pages` model, which doesn’t support a nested middleware model, but does support a single, root middleware instead.

Reviewer: ```suggestion
- **Faster execution with lower latency**: With nested Middleware, a single request could invoke multiple Middleware functions. A single Middleware means a single function execution, which is more efficient.
- **Less expensive**: Middleware usage is billed per invocation. Using nested Middleware, a single request could invoke multiple Middleware functions, meaning multiple Middleware charges per request. A single Middleware means a single invocation per request and is more cost effective.
- **Middleware can conveniently filter on things besides routes**: With nested Middleware, the Middleware files were located in the `pages` directory and Middleware was executed based on request paths. By moving to a single root Middleware, you can still execute code based on request paths, but you can now more conveniently execute Middleware based on other conditions, like `cookies` or the presence of a request header.
- **Deterministic execution ordering**: With nested Middleware, a single request could match multiple Middleware functions. For example, a request to `/dashboard/users/*` would invoke Middleware defined in both `/dashboard/users/_middleware.ts` and `/dashboard/_middleware.js`. However, the execution order is not easy to reason about. Moving to a single, root Middleware more explicitly defines execution order.
- **Supports Next.js Layouts (RFC)**: Moving to a single, root mMddleware helps support the new [Layouts (RFC) in Next.js](https://nextjs.org/blog/layouts-rfc).
```

### Explanation

Previously, you could create a `_middleware.ts` file under the `pages` directory at any level. Middleware execution was based on the file path where it was created.

Based on customer feedback, we have replaced this API with a single root Middleware, which provides the following improvements:

- **Faster execution with lower latency**  

>   With nested middleware, a single request could invoke multiple middleware functions. A single middleware means a single function execution, which is more efficient.

Reviewer: Can we put these like:

- **Thing:** description

Author: Done. 👍

### Explanation

> To respect the differences in client-side and server-side navigation, and to help ensure that developers do not build insecure middleware, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).

Reviewer: ```suggestion
To respect the differences in client-side and server-side navigation, and to help ensure that developers do not build insecure Middleware, we are removing the ability to send response bodies in Middleware. This ensures that Middleware is only used to `rewrite`, `redirect`, or modify the incoming request (e.g. [setting cookies](#cookies-api-revamped)).
```
